// angular2-tooltips/lib/tooltip.component.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var TooltipComponent = (function () {
    function TooltipComponent(elRef) {
        this.elRef = elRef;
        this.active = false;
    }
    TooltipComponent.prototype.renderTooltip = function (el, options) {
        this._relativeElement = el;
        this.tooltipClass = options.tooltipClass;
        if (options.content) {
            this._setTooltipContent(options.content);
            this._setTooltipPosition();
        }
        if (options.active) {
            this.active = true;
        }
        else {
            this.active = false;
        }
        return this.elRef;
    };
    TooltipComponent.prototype._calculateTooltipPosition = function () {
        var tooltipEl = this.elRef.nativeElement.querySelector('.tooltip');
        var relEl = this._relativeElement.nativeElement;
        var tooltipDimensions = [tooltipEl.offsetWidth, tooltipEl.offsetHeight];
        var relDimensions = [
            relEl.offsetLeft + relEl.offsetWidth / 2, relEl.offsetTop
        ];
        var targetX = (relDimensions[0] - tooltipDimensions[0] / 2);
        var adjustedX = Math.min(window.innerWidth - tooltipEl.offsetWidth, Math.max(4, targetX));
        var adjustedOffset = Math.max(Math.min((tooltipEl.offsetWidth / 2) - 10, targetX - adjustedX), -(tooltipEl.offsetWidth / 2) + 10);
        return {
            left: adjustedX,
            top: relDimensions[1] - tooltipDimensions[1],
            offset: adjustedOffset
        };
    };
    TooltipComponent.prototype._setTooltipContent = function (content) {
        var tooltipContentEl = this.elRef.nativeElement.querySelector('.tooltip-content');
        tooltipContentEl.innerHTML = content;
    };
    TooltipComponent.prototype._setPointerOffset = function (offset) {
        var tooltipArrow = this.elRef.nativeElement.querySelector('.tooltip-arrow');
        var translate = 'translateX(' + offset + 'px)';
        tooltipArrow.style.webkitTransform = translate;
        tooltipArrow.style.MozTransform = translate;
        tooltipArrow.style.msTransform = translate;
        tooltipArrow.style.OTransform = translate;
        tooltipArrow.style.transform = translate;
    };
    TooltipComponent.prototype._setTooltipPosition = function () {
        var tooltipEl = this.elRef.nativeElement.querySelector('.tooltip');
        var positions = this._calculateTooltipPosition();
        tooltipEl.style.left = positions.left + "px";
        tooltipEl.style.top = positions.top + "px";
        if (positions.offset != 0) {
            this._setPointerOffset(positions.offset);
        }
    };
    return TooltipComponent;
}());
TooltipComponent = __decorate([
    core_1.Component({
        template: "<div\n      class=\"tooltip\"\n      [@tooltipState]=\"active?'active':'inactive'\"\n      [ngClass]=\"tooltipClass\"\n    >\n      <span class=\"tooltip-content\"></span>\n      <span class=\"tooltip-arrow\"></span>\n    </div>\n  ",
        styles: ["\n    :host {\n      position:absolute;\n      background: rgba(0,0,0,0.8);\n      color: #fff;\n      font-size:1.2rem;\n      line-height: 1.2;\n      padding:8px 12px;\n      border-radius: 3px;\n    }\n    .tooltip-content {\n      min-width:100px;\n    }\n    .tooltip-arrow {\n      position:absolute;\n      bottom: -10px;\n      border: 5px solid transparent;\n      border-top-color: rgba(0,0,0,0.8);\n      left:50%;\n      margin-left:-5px;\n    }\n\n  "],
        animations: [
            core_1.trigger('tooltipState', [
                core_1.state('active', core_1.style({
                    opacity: 1,
                    transform: 'translate(0, -4px)'
                })),
                core_1.state('inactive', core_1.style({
                    opacity: 0,
                    transform: 'translate(0, 0)'
                })),
                core_1.transition(':enter', [
                    core_1.style({ opacity: 0, transform: 'translate(0,0)' }),
                    core_1.animate(300)
                ]),
                core_1.transition(':leave', [
                    core_1.style({ opacity: 0, transform: 'translate(0,0)' }),
                    core_1.animate(300)
                ]),
                core_1.transition('inactive => active', core_1.animate(300)),
                core_1.transition('active => inactive', core_1.animate(300))
            ])
        ]
    }),
    __metadata("design:paramtypes", [core_1.ElementRef])
], TooltipComponent);
exports.TooltipComponent = TooltipComponent;
